### DISCUSSION TITLE
What's the plan for byte-buddy-agent once dynamic agents are disallowed by default?

### DISCUSSION THREAD

**Discussion by ByteZ1337** [raphw/byte-buddy#1486] [General]
The OpenJDK team is currently planning to disallow the dynamic loading of Java Agents by default, as shown in [JEP 451](https://openjdk.org/jeps/451) and [JDK-8306275](https://bugs.openjdk.org/browse/JDK-8306275). This change would require the `-XX:-EnableDynamicAgentLoading` JVM flag to re-enable the attach mechanism.  
Will byte-buddy circumvent these new restrictions automatically or require the `EnableDynamicAgentLoading` flag to be manually set by end users?
Let me know if you need more information about our use-case.

**Comment by raphw** [comment]
Nothing in particular as Byte Buddy cannot work around such ultimate JVM restrictions. For testing, it will not be an issue as Maven/Gradle can add the parameter and there will be a sufficient exception explaining the deed, but other than that this simply disables the JVM infrastructure for it, so there will not be a work around.

**Reply by ByteZ1337** [reply]
>Nothing in particular as Byte Buddy cannot work around such ultimate JVM restrictions

There definitely are ways to work around these restrictions.
I was able to set the `EnableDynamicAgentLoading` flag during runtime on OpenJDK 20 in plain Java using the new [Project Panama memaccess API](https://github.com/openjdk/panama-foreign/blob/foreign-memaccess%2Babi/doc/panama_memaccess.md). xxDark was even able to do it from [another process](https://github.com/xxDark/Unsolver-8306275).

>but other than that this simply disables the JVM infrastructure for it

From what I can tell, the flag is only checked [here](https://github.com/openjdk/jdk/blob/9e4fc568a6f1a93c84a84d6cc5220c6eb4e546a5/src/hotspot/share/prims/jvmtiAgent.cpp#L491).

**Comment by raphw** [comment]
I thought about writing to native memory, but I would not expect this to work for a long time as there is a certain desire at OpenJDK to prohibit this. I will look into it.

**Reply by ByteZ1337** [reply]
I can send you my code to set the flag later if you want. Although It's pretty much the same as xxDark's code, just a bit more structured and using Panama instead of JNA. But since byte-buddy is already distributing natives anyway, maybe running everything through JNI would be the cleanest and fastest solution.

**Comment by nineninesevenfour** [comment]
Sorry, I overlooked this thread (otherwise would have maybe added it here), I have posted a how-to for test runs in discussion #1535

**Reply by ByteZ1337** [reply]
Sadly not applicable in our use case

**Comment by raphw** [comment]
I played with an approach to override this, but it requires special system privileges, what is even harder to set up, so its not really feasible.
